dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp

```diff
@@ -121,15 +142,71 @@ bool PredicateExpressionsOptimizer::allowPushDown(const ASTSelectQuery * subquer
         {
             const auto & function = FunctionFactory::instance().tryGet(subquery_function->name, context);
 
-            /// Skip lambda、tuple and other special functions
+            /// Skip lambda, tuple and other special functions
             if (function && function->isStateful())
                 return false;
         }
+    }
+
+    const auto * ast_join = ast_select->join();
+
+    /// XXX: code depends on the order and availability of AST children
+    const auto * left_table_expr = ast_join ? ast_select->tables->as<ASTTablesInSelectQuery>()
```

Comment:
Is it better to set pointers with nullptrs first & make common if (ast_join) for both of them?

Reply:
Refactored